### PR TITLE
BUG: Fix segfault in lib.isnullobj

### LINKED ISSUE #13717: Seqfault on creation of dataframe with np.empty_like
This code segfaults:
```python
import numpy as np
import pandas as pd
df = pd.DataFrame(np.random.randn(3, 3), index='A B C'.split())
pd.DataFrame(np.empty_like(df.index))
```

INSTALLED VERSIONS
------------------
commit: None
python: 3.5.2.final.0
python-bits: 64
OS: Linux
OS-release: 3.13.0-91-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.18.1
nose: 1.3.7
pip: 8.1.2
setuptools: 23.0.0
Cython: 0.24
numpy: 1.11.1
scipy: 0.17.1
statsmodels: 0.6.1
xarray: 0.7.2
IPython: 5.0.0
sphinx: None
patsy: 0.4.1
dateutil: 2.3
pytz: 2016.4
blosc: None
bottleneck: 1.1.0
tables: 3.2.2
numexpr: 2.6.0
matplotlib: 1.5.1
openpyxl: None
xlrd: None
xlwt: None
xlsxwriter: None
lxml: None
bs4: 4.4.1
html5lib: 0.999
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
jinja2: 2.8
boto: None
pandas_datareader: None

### PR DESCRIPTION
**Pull request by gfyoung**
Weird segfault arises when you call `lib.isnullobj` with an array that uses 0-field values to mean `None`.
Changed input to be a `Python` object (i.e. no typing), and the segfault went away.  Discovered when there were segfaults in printing a `DataFrame` containing such an array.

Closes #13717.

### CODE REVIEW

File: pandas/lib.pyx [hunk 1]
```diff
@@ -342,7 +342,7 @@ def item_from_zerodim(object val):
 
 @cython.wraparound(False)
 @cython.boundscheck(False)
-def isnullobj(ndarray[object] arr):
+def isnullobj(arr):
```
Comment (jreback) [inline pandas/lib.pyx]: there is another routine isnull right below this

**Comment by jreback** [comment]
you need to check if it's iterable now eg s scalar will fail

**Comment by jreback** [comment]
or at least assert that (as this cannot be called with s scalar)

File: pandas/tests/test_lib.py [hunk 2]
```diff
@@ -184,6 +185,52 @@ def test_get_reverse_indexer(self):
...
+
+    def test_basic(self):
+        arr = np.array([1, None, 'foo', -5.1, pd.NaT, np.nan])
+        expected = np.array([False, True, False, False, True, True])
+
+        result = lib.isnullobj(arr)
+        tm.assert_numpy_array_equal(result, expected)
+
+    def test_non_ndarray_inp(self):
+        ind = pd.Index([1, None, 'foo', -5.1, pd.NaT, np.nan])
```
Comment (jreback) [inline pandas/tests/test_lib.py]: put all of these tests with the null tests - you will have to find them

Comment (gfyoung) [inline pandas/tests/test_lib.py]: Not a very helpful comment tbh.  This a `lib` method, so I don't see why we don't put the tests here and instead go "hunting" for unit tests like this.

Comment (jreback) [inline pandas/tests/test_lib.py]: because you will now have tests in multiple places for the same thing 

some lib tests see segrated

try types/missing

Comment (gfyoung) [inline pandas/tests/test_lib.py]: But they're unit tests for two different methods, one of which is a helper for another.  Sure they might be similar, but from an organisation perspective, IMO it's easier to put the tests next to the module whose methods / functions you are testing.  That is an easier way of searching AND an easier way of organising.

Comment (jreback) [inline pandas/tests/test_lib.py]: the bottom line is to be logical where test go do they r easy to fine

my point is that tests for this should be in 1 place 
u can pick but consolidate the like tests 

or these functions not tested anywhere else??

Comment (gfyoung) [inline pandas/tests/test_lib.py]: The specific `lib` functions aren't being tested anywhere, so I intended to put unit tests for them in `test_lib.py`.  The `null` functions you were referring to are those exposed at the Python layer, not the internal ones (which is what I am fixing) that are written in Cython.

**Comment by gfyoung** [comment]
@jreback : good point about the scalar thing - I think I actually don't need to remove `ndarray` entirely.  If I remove the `object` dtype specification, that might be sufficient.  Testing it right now.

File: pandas/lib.pyx (hunk 1 above)
Comment (gfyoung) [inline pandas/lib.pyx]: Sigh...will fix.

**Comment by gfyoung** [comment]
@jreback : "fixed" all four `isnullobj` methods and added tests for all of them, and Travis is still passing.  Ready to merge if there are no other concerns.

**Comment by jreback** [comment]
lgtm. can you give a perf test and see if anything material (I don't think so, but can't hurt). as this is the null tester for strings.

**Comment by gfyoung** [comment]
What do you mean `null` tester for strings?  What method are you referring to?

**Comment by jreback** [comment]
this is called for null testing on strings

**Comment by gfyoung** [comment]
From where exactly?  Are you talking about the `pandas.types.missing` methods (also, why focus on strings specifically?)

**Comment by jreback** [comment]
look at how isnull works. this is the only 1 of 2 places this is called.

**Comment by gfyoung** [comment]
Okay, that's what I thought.  Where should this perf test go in `asv_bench`?  In `strings.py` I presume?

**Comment by jreback** [comment]
oh, i mean there prob is a benchmark that tells null on strings already, if not see where the null tests are (I *think* they are all in one place).

**Comment by gfyoung** [comment]
@jreback : I found `isnull` benchmarks in the `DataFrame` benchmarks, so I added one for strings there.

**Comment by gfyoung** [comment]
Travis is doing weird things.  Putting `[ci skip]` in the commit until it corrects itself.

File: asv_bench/benchmarks/frame_methods.py [hunk 3]
```diff
@@ -606,6 +607,21 @@ def time_frame_isnull(self):
...
 
+class frame_isnull_strings(object):
+    goal_time = 0.2
+
+    def setup(self):
+        np.random.seed(1234)
+        self.sample = np.array(list(string.ascii_lowercase) +
+                               list(string.ascii_uppercase) +
+                               list(string.whitespace))
+        self.data = np.random.choice(self.choice, (1000, 1000))
```
Comment (jreback) [inline asv_bench/benchmarks/frame_methods.py]: pls test, this was not building on asv (choice -> sample)

**Comment by jreback** [comment]
thanks

minor change I made in the asv
also it IS slightly slower, so I guess the emitted c-code is slightly less optimal w/o the object typing (but can't have seg faults of course :)

File: asv_bench/benchmarks/frame_methods.py (hunk 3 above)
Comment (gfyoung) [inline asv_bench/benchmarks/frame_methods.py]: Whoops, sorry!  Was in a bit of a rush when I wrote it.

**Pull request closed without merging**